### PR TITLE
Remove inaccuracies and improve experience with ML.NET transfer learning tutorial

### PR DESCRIPTION
**Pull request by natke**
Preview link: https://review.docs.microsoft.com/en-us/dotnet/machine-learning/tutorials/image-classification?branch=pr-en-us-14531

[AB#1596621](https://dev.azure.com/mseng/677da0fb-b067-4f77-b89b-f32c12bb8617/_workitems/edit/1596621)

### CODE REVIEW

File: docs/machine-learning/tutorials/image-classification.md [hunk 1]
```diff
@@ -111,336 +111,249 @@ toaster2.png	appliance
...
+    [!code-csharp[LoadImageData](../../../samples/machine-learning/tutorials/TransferLearningTF/Program.cs#LoadImageData)]
 
-[!code-csharp[LoadData](../../../samples/machine-learning/tutorials/TransferLearningTF/Program.cs#LoadData "Load the data")]
+1. Make a single prediction, by adding the following code as the next line in the `ClassifySingleImage` method:
 
-### Extract Features and transform the data
+    [!code-csharp[PredictSingle](../../../samples/machine-learning/tutorials/TransferLearningTF/Program.cs#PredictSingle)]
 
-Pre-processing and cleaning data are important tasks that occur before a dataset is used effectively for machine learning.  Using data without these modeling tasks can produce misleading results.
+    The [PredictionEngine class](xref:Microsoft.ML.PredictionEngine%602) is a convenience API that performs a prediction on a single instance of data. The [Predict()](xref:Microsoft.ML.PredictionEngine%602.Predict%2A) function makes a prediction on a single column of data.
```
Comment (luisquintanilla) [inline docs/machine-learning/tutorials/image-classification.md]: Suggestion:

The [PredictionEngine class](xref:Microsoft.ML.PredictionEngine%602) is a convenience API that performs a prediction on a single instance of data. To get the prediction, use the [Predict()](xref:Microsoft.ML.PredictionEngine%602.Predict%2A) method.

File: docs/machine-learning/tutorials/image-classification.md [hunk 2]
```diff
@@ -111,336 +111,249 @@ toaster2.png	appliance
...
 
-Machine learning algorithms understand [featurized](../resources/glossary.md#feature) data, and when dealing with deep neural networks you must adapt the images to the format expected by the network. That format is a [numeric vector](../resources/glossary.md#numerical-feature-vector).
+1. Display the prediction result as the next line of code in the `ClassifySingleImage()` method:
 
-After training and evaluation, predict with the **Label** column values. As you're using a pre-trained model, map fields to the new model with the [MapValueToKey()](xref:Microsoft.ML.ConversionsExtensionsCatalog.MapValueToKey%2A) method. This method transforms the `Label` into a numeric key type (`LabelTokey`) column and add it as new dataset column:  Name this `estimator` as you'll also add the trainer to it. Add the next line of code:
+   [!code-csharp[DisplayPrediction](../../../samples/machine-learning/tutorials/TransferLearningTF/Program.cs#DisplayPrediction)]
 
-[!code-csharp[MapValueToKey1](../../../samples/machine-learning/tutorials/TransferLearningTF/Program.cs#MapValueToKey1)]
+1. Add a method to generate the model
```
Comment (luisquintanilla) [inline docs/machine-learning/tutorials/image-classification.md]: I would move this to the **Construct the ML.NET model pipeline** section.

File: docs/machine-learning/tutorials/image-classification.md [hunk 3]
```diff
@@ -111,336 +111,249 @@ toaster2.png	appliance
...
 
-To add the training algorithm, call the `mlContext.MulticlassClassification.Trainers.LbfgsMaximumEntropy()` wrapper method.  The [LbfgsMaximumEntropy](xref:Microsoft.ML.Trainers.LbfgsMaximumEntropyMulticlassTrainer) is appended to the `estimator` and accepts the Inception image features (`softmax2_pre_activation`) and the `Label` input parameters to learn from the historic data.  Add the trainer with the following code:
+1. Add the estimator to load the TensorFlow model, and score it:
 
-[!code-csharp[AddTrainer](../../../samples/machine-learning/tutorials/TransferLearningTF/Program.cs#AddTrainer)]
+    [!code-csharp[ScoreTensorFlowModel](../../../samples/machine-learning/tutorials/TransferLearningTF/Program.cs#ScoreTensorFlowModel)]
 
-You also need to map the `predictedlabel` to the `predictedlabelvalue`:
+    The next stage in the pipeline is to load the TensorFlow model into memory, then process the vector of pixel values through the TensorFlow model network. Applying inputs to a deep learning model is referred to as **Scoring**. When using the model in its entirety, scoring makes an inference, or prediction.
```
Comment (luisquintanilla) [inline docs/machine-learning/tutorials/image-classification.md]: I would reword the last two sentences to clarify scoring as the process of getting a prediction, instead of applying inputs.

File: docs/machine-learning/tutorials/image-classification.md [hunk 4]
```diff
@@ -111,336 +111,249 @@ toaster2.png	appliance
...
+1. Add the estimator to load the TensorFlow model, and score it:
 
-[!code-csharp[AddTrainer](../../../samples/machine-learning/tutorials/TransferLearningTF/Program.cs#AddTrainer)]
+    [!code-csharp[ScoreTensorFlowModel](../../../samples/machine-learning/tutorials/TransferLearningTF/Program.cs#ScoreTensorFlowModel)]
 
-You also need to map the `predictedlabel` to the `predictedlabelvalue`:
+    The next stage in the pipeline is to load the TensorFlow model into memory, then process the vector of pixel values through the TensorFlow model network. Applying inputs to a deep learning model is referred to as **Scoring**. When using the model in its entirety, scoring makes an inference, or prediction. 
 
-[!code-csharp[MapValueToKey2](../../../samples/machine-learning/tutorials/TransferLearningTF/Program.cs#MapValueToKey2)]
+    In this case, you use all of the TensorFlow model except the last layer, which is the layer that makes the inference. The output of the penultimate layer is labeled `softmax_2_preactivation`. This output of this layer is effectively a vector of features that characterize the original input images.
```
Comment (luisquintanilla) [inline docs/machine-learning/tutorials/image-classification.md]:
```suggestion
    In this case, you use all of the TensorFlow model except the last layer, which is the layer that makes the inference. The output of the penultimate layer is labeled `softmax_2_preactivation`. The output of this layer is effectively a vector of features that characterize the original input images.
```

File: docs/machine-learning/tutorials/image-classification.md [hunk 5]
```diff
@@ -111,336 +111,249 @@ toaster2.png	appliance
...
+    This feature vector generated by the TensorFlow model will be used as input to the ML.NET trainer below.
 
-[!code-csharp[TrainModel](../../../samples/machine-learning/tutorials/TransferLearningTF/Program.cs#TrainModel)]
+1. Add the estimator to map the string labels in the training data to integer key values:
 
-The [Transform()](xref:Microsoft.ML.ITransformer.Transform%2A) method makes predictions for multiple provided input rows of a test dataset.  Transform the `Training` data by adding the following code to `ReuseAndTuneInceptionModel()`:
+    [!code-csharp[MapValueToKey](../../../samples/machine-learning/tutorials/TransferLearningTF/Program.cs#MapValueToKey)]
 
-[!code-csharp[TransformData](../../../samples/machine-learning/tutorials/TransferLearningTF/Program.cs#TransformData)]
+    The ML.NET trainer that is appended next requires its labels to be in `key` format rather than arbitrary strings. A key is a number that has a one to one mapping to a string value.
```
Comment (luisquintanilla) [inline docs/machine-learning/tutorials/image-classification.md]: Suggestion: Add xref https://docs.microsoft.com/en-us/dotnet/api/microsoft.ml.data.keydataviewtype?view=ml-dotnet for `KeyDataViewType`

File: docs/machine-learning/tutorials/image-classification.md [hunk 6]
```diff
@@ -111,336 +111,249 @@ toaster2.png	appliance
...
 
-You also need to map the `predictedlabel` to the `predictedlabelvalue`:
+    The next stage in the pipeline is to load the TensorFlow model into memory, then process the vector of pixel values through the TensorFlow model network. Applying inputs to a deep learning model is referred to as **Scoring**. When using the model in its entirety, scoring makes an inference, or prediction. 
 
-[!code-csharp[MapValueToKey2](../../../samples/machine-learning/tutorials/TransferLearningTF/Program.cs#MapValueToKey2)]
+    In this case, you use all of the TensorFlow model except the last layer, which is the layer that makes the inference. The output of the penultimate layer is labeled `softmax_2_preactivation`. This output of this layer is effectively a vector of features that characterize the original input images.
 
-The `Fit()` method trains your model by transforming the dataset and applying the training. Fit the model to the training dataset and return the trained model by adding the following as the next line of code in the `ReuseAndTuneInceptionModel()` method:
+    This feature vector generated by the TensorFlow model will be used as input to the ML.NET trainer below.
```
Comment (luisquintanilla) [inline docs/machine-learning/tutorials/image-classification.md]: Suggestion: The feature vector generated by the TensorFlow model is used as input to an ML.NET trainer.

Question: Is it a trainer or estimator? Especially since it's the value-to-key mapping and not the algorithm.

File: docs/machine-learning/tutorials/image-classification.md [hunk 7]
```diff
@@ -111,336 +111,249 @@ toaster2.png	appliance
...
+1. Add the estimator to map the predicted key value back into a string:
 
-[!code-csharp[CallDisplayResults1](../../../samples/machine-learning/tutorials/TransferLearningTF/Program.cs#CallDisplayResults1)]
+    [!code-csharp[MapKeyToValue](../../../samples/machine-learning/tutorials/TransferLearningTF/Program.cs#MapKeyToValue)]
 
-Once you have the prediction set, the [Evaluate()](xref:Microsoft.ML.RecommendationCatalog.Evaluate%2A) method:
+## Train the model
 
-* Assesses the model (compares the predicted values with the actual dataset `Labels`).
+1. Load the training data using the `MLContext.Data.LoadFromTextFile` wrapper. Add the following code as the next line in the `GenerateModel()` method:
```
Comment (luisquintanilla) [inline docs/machine-learning/tutorials/image-classification.md]: Add the xref for `LoadFromTextFile` https://docs.microsoft.com/en-us/dotnet/api/microsoft.ml.textloadersavercatalog.loadfromtextfile?view=ml-dotnet#Microsoft_ML_TextLoaderSaverCatalog_LoadFromTextFile_Microsoft_ML_DataOperationsCatalog_System_String_Microsoft_ML_Data_TextLoader_Options_

File: docs/machine-learning/tutorials/image-classification.md [hunk 8]
```diff
@@ -111,336 +111,249 @@ toaster2.png	appliance
...
+    Data in ML.NET is represented as an [IDataView class](xref:Microsoft.ML.IDataView). `IDataView` is a flexible, efficient way of describing tabular data (numeric and text). Data can be loaded from a text file or in real time (for example, SQL database or log files) to an `IDataView` object.
 
-[!code-csharp[Evaluate](../../../samples/machine-learning/tutorials/TransferLearningTF/Program.cs#Evaluate)]
+1. Train the model with the data loaded above:
 
-The following metrics are evaluated for image classification:
+    [!code-csharp[TrainModel](../../../samples/machine-learning/tutorials/TransferLearningTF/Program.cs#TrainModel)]
 
-* `Log-loss` - see [Log Loss](../resources/glossary.md#log-loss). You want Log-loss to be as close to zero as possible.
+    The `Fit()` method trains your model by applying the training dataset to the pipeline.
```
Comment (luisquintanilla) [inline docs/machine-learning/tutorials/image-classification.md]:
```suggestion
    The `Fit()` method trains your model by applying the pipeline to the training dataset.
```

File: docs/machine-learning/tutorials/image-classification.md [hunk 9]
```diff
@@ -111,336 +111,249 @@ toaster2.png	appliance
...
+## Evaluate the accuracy of the model
 
-Use the following code to display the metrics, share the results, and then act on them:
+1. Load and transform the test data, by adding the following code to the next line of the `GenerateModel` method:
 
-[!code-csharp[DisplayMetrics](../../../samples/machine-learning/tutorials/TransferLearningTF/Program.cs#DisplayMetrics)]
+    [!code-csharp[LoadAndTransformTestData](../../../samples/machine-learning/tutorials/TransferLearningTF/Program.cs#LoadAndTransformTestData "Load and transform test data")]
 
- Add the following code to return the trained model as the next line:
+    There are a few sample images that you can use to evaluate the model. Like the training data, these need to be loaded into an IDataView, so that they can be transformed by the model.
```
Comment (luisquintanilla) [inline docs/machine-learning/tutorials/image-classification.md]:
```suggestion
    There are a few sample images that you can use to evaluate the model. Like the training data, these need to be loaded into an `IDataView`, so that they can be transformed by the model.
```

File: docs/machine-learning/tutorials/image-classification.md [hunk 10]
```diff
@@ -111,336 +111,249 @@ toaster2.png	appliance
...
-```csharp
-public static void ClassifySingleImage(MLContext mlContext, string imagePath, string outputModelLocation, ITransformer model)
-{
-
-}
-```
+## Tie it all together 
 
-First, create an `ImageData` class that contains the fully qualified path and image file name for the single `ImagePath`. Add the following code as the next  lines in the `ClassifySingleImage()` method:
+1. Add the following code to return the trained model as the next line:
```
Comment (luisquintanilla) [inline docs/machine-learning/tutorials/image-classification.md]: I would move this step to the previous section to complete the `GenerateModel` method.

File: docs/machine-learning/tutorials/image-classification.md [hunk 11]
```diff
@@ -111,336 +111,249 @@ toaster2.png	appliance
...
-
-Create the `ClassifySingleImage()` method, just after the `ClassifyImages()` method and just before the `PairAndDisplayResults()` method, using the following code:
-
-```csharp
-public static void ClassifySingleImage(MLContext mlContext, string imagePath, string outputModelLocation, ITransformer model)
-{
-
-}
-```
+## Tie it all together
```
Comment (luisquintanilla) [inline docs/machine-learning/tutorials/image-classification.md]: Rework title. Something like Use the model or Run the application and include the Results as a subsection.

File: docs/machine-learning/tutorials/image-classification.md [hunk 12]
```diff
@@ -1,62 +1,64 @@
 ---
-title: 'Tutorial: Retrain TensorFlow image classifier - transfer learning'
-description: Learn how to retrain an image classification TensorFlow model with transfer learning and ML.NET. The original model was trained to classify individual images. After retraining, the new model organizes the images into broad categories.
-ms.date: 07/09/2019
+title: 'Tutorial: Generate an ML.NET image classification model from a pre-trained TensorFlow model'
+description: Learn how to transfer the intelligence from an existing TensorFlow model into a new ML.NET image classification model. The TensorFlow model was trained to classify images into a thousand categories. The ML.NET model makes use of transfer learning to classify images into fewer broader categories.
```
Comment (luisquintanilla) [inline docs/machine-learning/tutorials/image-classification.md]:
```suggestion
description: Learn how to transfer the knowledge from an existing TensorFlow model into a new ML.NET image classification model. The TensorFlow model was trained to classify images into a thousand categories. The ML.NET model makes use of transfer learning to classify images into fewer broader categories.
```

File: docs/machine-learning/tutorials/image-classification.md [hunk 13]
```diff
@@ -1,62 +1,64 @@
...
+ms.date: 09/26/2019
 ms.topic: tutorial
 ms.custom: mvc, title-hack-0612
-#Customer intent: As a developer, I want to reuse a pre-trained TensorFlow model with ML.NET so that I can classify images with a small amount of training data.
+#Customer intent: As a developer, I want to use a pre-trained TensorFlow model with ML.NET so that I can classify images with a small amount of training data.
```
Comment (luisquintanilla) [inline docs/machine-learning/tutorials/image-classification.md]: author metadata missing

**Review by luisquintanilla — COMMENTED** [review]
Looks good. Added a few comments.

File: docs/machine-learning/tutorials/image-classification.md [hunk 14]
```diff
@@ -1,62 +1,64 @@
...
 ---
-# Tutorial: Retrain a TensorFlow image classifier with transfer learning and ML.NET
+# Tutorial: Generate an ML.NET image classification model from a pre-trained TensorFlow model
 
-Learn how to retrain an image classification TensorFlow model with transfer learning and ML.NET. The original model was trained to classify individual images. After retraining, the new model organizes the images into broad categories. 
+Learn how to transfer the intelligence from an existing TensorFlow model into a new ML.NET image classification model.
```
[the remark this answers is not in the document]
Comment (natke) [reply docs/machine-learning/tutorials/image-classification.md]: I thought about this for a long time! Is it really knowledge? :-)  Maybe I should say artificial intelligence! I'll change it

File: docs/machine-learning/tutorials/image-classification.md [hunk 15]
```diff
@@ -111,336 +111,249 @@ toaster2.png	appliance
...
-}
-```
-
-The `Transform()` method populated `ImagePath` in `ImagePrediction` along with the predicted fields. As the ML.NET process progresses, each component adds columns, and this makes it easy to display the results:
+    [!code-csharp[DisplayPredictions](../../../samples/machine-learning/tutorials/TransferLearningTF/Program.cs#DisplayPredictions)]
 
-[!code-csharp[DisplayPredictions](../../../samples/machine-learning/tutorials/TransferLearningTF/Program.cs#DisplayPredictions)]
-
-You'll call the `DisplayResults()` method in the two image classification methods.
+    The `Transform()` method populated `ImagePath` in `ImagePrediction` along with the predicted fields. As the ML.NET process progresses, each component adds columns, and this makes it easy to display the results:
```
[the remark this answers is not in the document]
Comment (natke) [reply docs/machine-learning/tutorials/image-classification.md]: Well spotted. That was a hangover from the days when the training data was being used to verify the model :) I've removed it

File: docs/machine-learning/tutorials/image-classification.md [hunk 16]
```diff
@@ -111,336 +111,249 @@ toaster2.png	appliance
...
-
-```csharp
-public static IEnumerable<ImageData> ReadFromTsv(string file, string folder)
-{
-
-}
-```
-
-The following code parses through the `tags.tsv` file to add the file path to the image file name for the `ImagePath` property and load it and the `Label` into an `ImageData` object. Add it as the first line of the `ReadFromTsv()` method.  You need the fully qualified file path to display the prediction results.
+1. Create the `ReadFromTsv()` method, just after the `PairAndDisplayResults()` method, using the following code:
```
[the remark this answers is not in the document]
Comment (natke) [reply docs/machine-learning/tutorials/image-classification.md]: Yep. Fixed

File: docs/machine-learning/tutorials/image-classification.md (hunk 1 above)
Comment (natke) [reply docs/machine-learning/tutorials/image-classification.md]: Changed

File: docs/machine-learning/tutorials/image-classification.md (hunk 2 above)
Comment (natke) [reply docs/machine-learning/tutorials/image-classification.md]: Yeah, that's where I originally had it. I'll move it back

File: docs/machine-learning/tutorials/image-classification.md (hunk 3 above)
Comment (natke) [reply docs/machine-learning/tutorials/image-classification.md]: Fixed.

File: docs/machine-learning/tutorials/image-classification.md (hunk 4 above)
Comment (natke) [reply docs/machine-learning/tutorials/image-classification.md]: Fixed

File: docs/machine-learning/tutorials/image-classification.md (hunk 6 above)
Comment (natke) [reply docs/machine-learning/tutorials/image-classification.md]: I hedged and changed it to ML.NET training algorithm

File: docs/machine-learning/tutorials/image-classification.md (hunk 7 above)
Comment (natke) [reply docs/machine-learning/tutorials/image-classification.md]: Fixed

File: docs/machine-learning/tutorials/image-classification.md (hunk 8 above)
Comment (natke) [reply docs/machine-learning/tutorials/image-classification.md]: OOh, I don't think I agree with that. Aren't you applying the data to the pipeline?

File: docs/machine-learning/tutorials/image-classification.md (hunk 9 above)
Comment (natke) [reply docs/machine-learning/tutorials/image-classification.md]: Fixed.

File: docs/machine-learning/tutorials/image-classification.md (hunk 11 above)
Comment (natke) [reply docs/machine-learning/tutorials/image-classification.md]: Good idea. Changed.

File: docs/machine-learning/tutorials/image-classification.md (hunk 10 above)
Comment (natke) [reply docs/machine-learning/tutorials/image-classification.md]: Changed

File: docs/machine-learning/tutorials/image-classification.md (hunk 12 above)
Comment (natke) [reply docs/machine-learning/tutorials/image-classification.md]: Changed

File: docs/machine-learning/tutorials/image-classification.md (hunk 13 above)
Comment (natke) [reply docs/machine-learning/tutorials/image-classification.md]: Fixed

File: docs/machine-learning/tutorials/image-classification.md (hunk 5 above)
Comment (natke) [reply docs/machine-learning/tutorials/image-classification.md]: Yeah, I thought about that. But that API documentation is not great atm.

**Review by luisquintanilla — APPROVED** [review]
Nice job! :shipit:
